// @ngxs/store/esm2015/src/internal/config-validator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { CONFIG_MESSAGES as MESSAGES, VALIDATION_CODE as CODE } from '../configs/messages.config';
import { NgxsConfig } from '../symbols';
import { HostEnvironment } from '../host-environment/host-environment';
import * as ɵngcc0 from '@angular/core';
export class ConfigValidator {
    /**
     * @param {?} _host
     * @param {?} _config
     */
    constructor(_host, _config) {
        this._host = _host;
        this._config = _config;
    }
    /**
     * @private
     * @return {?}
     */
    get isIncorrectProduction() {
        return !this._host.isDevMode() && this._config.developmentMode;
    }
    /**
     * @private
     * @return {?}
     */
    get isIncorrectDevelopment() {
        return this._host.isDevMode() && !this._config.developmentMode;
    }
    /**
     * @return {?}
     */
    verifyDevMode() {
        if (this._host.isTestMode()) {
            return;
        }
        if (this.isIncorrectProduction) {
            console.warn(MESSAGES[CODE.INCORRECT_PRODUCTION]());
        }
        else if (this.isIncorrectDevelopment) {
            console.warn(MESSAGES[CODE.INCORRECT_DEVELOPMENT]());
        }
    }
}
ConfigValidator.ɵfac = function ConfigValidator_Factory(t) { return new (t || ConfigValidator)(ɵngcc0.ɵɵinject(HostEnvironment), ɵngcc0.ɵɵinject(NgxsConfig)); };
ConfigValidator.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ConfigValidator, factory: ConfigValidator.ɵfac });
/** @nocollapse */
ConfigValidator.ctorParameters = () => [
    { type: HostEnvironment },
    { type: NgxsConfig }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ConfigValidator, [{
        type: Injectable
    }], function () { return [{ type: HostEnvironment }, { type: NgxsConfig }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    ConfigValidator.prototype._host;
    /**
     * @type {?}
     * @private
     */
    ConfigValidator.prototype._config;
}
