// @ngxs/store/esm5/src/host-environment/host-environment.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { NG_DEV_MODE, NG_TEST_MODE } from '../symbols';
import * as ɵngcc0 from '@angular/core';
var HostEnvironment = /** @class */ (function () {
    function HostEnvironment(isDevMode, isTestMode) {
        this.isDevMode = isDevMode;
        this.isTestMode = isTestMode;
    }
    /** @nocollapse */
    HostEnvironment.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NG_DEV_MODE,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [NG_TEST_MODE,] }] }
    ]; };
HostEnvironment.ɵfac = function HostEnvironment_Factory(t) { return new (t || HostEnvironment)(ɵngcc0.ɵɵinject(NG_DEV_MODE), ɵngcc0.ɵɵinject(NG_TEST_MODE)); };
HostEnvironment.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: HostEnvironment, factory: function (t) { return HostEnvironment.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(HostEnvironment, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [NG_DEV_MODE]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [NG_TEST_MODE]
            }] }]; }, null); })();
    return HostEnvironment;
}());
export { HostEnvironment };
if (false) {
    /** @type {?} */
    HostEnvironment.prototype.isDevMode;
    /** @type {?} */
    HostEnvironment.prototype.isTestMode;
}
