// @ngxs/router-plugin/esm5/src/router.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { NgxsModule } from '@ngxs/store';
import { RouterState } from './router.state';
import { DefaultRouterStateSerializer, RouterStateSerializer } from './serializer';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngxs/store';
var NgxsRouterPluginModule = /** @class */ (function () {
    function NgxsRouterPluginModule() {
    }
    /**
     * @return {?}
     */
    NgxsRouterPluginModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgxsRouterPluginModule,
            providers: [{ provide: RouterStateSerializer, useClass: DefaultRouterStateSerializer }]
        };
    };
NgxsRouterPluginModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxsRouterPluginModule });
NgxsRouterPluginModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxsRouterPluginModule_Factory(t) { return new (t || NgxsRouterPluginModule)(); }, imports: [[NgxsModule.forFeature([RouterState])]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxsRouterPluginModule, { imports: [ɵngcc1.ɵbc] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxsRouterPluginModule, [{
        type: NgModule,
        args: [{
                imports: [NgxsModule.forFeature([RouterState])]
            }]
    }], function () { return []; }, null); })();
    return NgxsRouterPluginModule;
}());
export { NgxsRouterPluginModule };
