// ngxs-reset-plugin/esm2015/lib/reset.handler.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/reset.handler.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Actions, InitState, ofActionSuccessful, Store, UpdateState } from '@ngxs/store';
import { take } from 'rxjs/operators';
import { ResetService } from './reset.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngxs/store';
export class ResetHandler {
    /**
     * @param {?} actions$
     * @param {?} store
     * @param {?} resetService
     */
    constructor(actions$, store, resetService) {
        this.actions$ = actions$;
        this.store = store;
        this.resetService = resetService;
        this.actions$
            .pipe(ofActionSuccessful(InitState), take(1))
            .subscribe((/**
         * @return {?}
         */
        () => (this.resetService.initialState = this.store.snapshot())));
        this.actions$.pipe(ofActionSuccessful(UpdateState)).subscribe((/**
         * @param {?} __0
         * @return {?}
         */
        ({ addedStates }) => (this.resetService.initialState = Object.assign(Object.assign({}, this.resetService.initialState), addedStates))));
    }
}
ResetHandler.ɵfac = function ResetHandler_Factory(t) { return new (t || ResetHandler)(ɵngcc0.ɵɵinject(ɵngcc1.Actions), ɵngcc0.ɵɵinject(ɵngcc1.Store), ɵngcc0.ɵɵinject(ResetService)); };
ResetHandler.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: ResetHandler, factory: ResetHandler.ɵfac });
/** @nocollapse */
ResetHandler.ctorParameters = () => [
    { type: Actions },
    { type: Store },
    { type: ResetService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ResetHandler, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.Actions }, { type: ɵngcc1.Store }, { type: ResetService }]; }, null); })();
if (false) {
    /**
     * @type {?}
     * @private
     */
    ResetHandler.prototype.actions$;
    /**
     * @type {?}
     * @private
     */
    ResetHandler.prototype.store;
    /**
     * @type {?}
     * @private
     */
    ResetHandler.prototype.resetService;
}
