// @ngxs/storage-plugin/esm5/src/storage.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, PLATFORM_ID, InjectionToken } from '@angular/core';
import { NGXS_PLUGINS } from '@ngxs/store';
import { STORAGE_ENGINE, NGXS_STORAGE_PLUGIN_OPTIONS } from './symbols';
import { NgxsStoragePlugin } from './storage.plugin';
import { storageOptionsFactory, engineFactory } from './internals';
/** @type {?} */
import * as ɵngcc0 from '@angular/core';
export var USER_OPTIONS = new InjectionToken('USER_OPTIONS');
var NgxsStoragePluginModule = /** @class */ (function () {
    function NgxsStoragePluginModule() {
    }
    /**
     * @param {?=} options
     * @return {?}
     */
    NgxsStoragePluginModule.forRoot = /**
     * @param {?=} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: NgxsStoragePluginModule,
            providers: [
                {
                    provide: NGXS_PLUGINS,
                    useClass: NgxsStoragePlugin,
                    multi: true
                },
                {
                    provide: USER_OPTIONS,
                    useValue: options
                },
                {
                    provide: NGXS_STORAGE_PLUGIN_OPTIONS,
                    useFactory: storageOptionsFactory,
                    deps: [USER_OPTIONS]
                },
                {
                    provide: STORAGE_ENGINE,
                    useFactory: engineFactory,
                    deps: [NGXS_STORAGE_PLUGIN_OPTIONS, PLATFORM_ID]
                }
            ]
        };
    };
NgxsStoragePluginModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxsStoragePluginModule });
NgxsStoragePluginModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxsStoragePluginModule_Factory(t) { return new (t || NgxsStoragePluginModule)(); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxsStoragePluginModule, [{
        type: NgModule
    }], function () { return []; }, null); })();
    return NgxsStoragePluginModule;
}());
export { NgxsStoragePluginModule };
